### ISSUE TITLE
[JOSS] Installation

### ISSUE THREAD

**Issue by willu47** [SESMG/SESMG#188]
Hi, this is a question regarding installtion from [JOSS Review #5519](https://github.com/openjournals/joss-reviews/issues/5519)

Please could you explain to me why you have decided against using an automated package management solution such as setuptools, conda or venv to handle the installation of dependencies for SESMG?

As a MacOS + conda user, I was unable to follow the installation instructions, as they make the assumption that I am installing Python from scratch on a systemwide basis.

I installed the package easily by using the following commands:

    git clone https://github.com/SESMG/SESMG.git
    conda create -n sesmg python=3.10 coincbc geos graphviz postgresql
    pip install -r requirements.txt

    mkdir files_deviant_operating_system
    mv Linux_installation.sh files_deviant_operating_system
    mv Run_SESMG_for_Linux.sh files_deviant_operating_system
    mv Windows_installation.cmd files_deviant_operating_system
    mv Run_SESMG_for_windows.cmd files_deviant_operating_system

    python -m streamlit run "program_files/GUI_st/1_Main_Application.py" osascript -e 'tell application "Terminal" to quit'

But I don't understand why you have you not packaged the SESMG as a Python package.  There's nothing in the above which could not be solved by a Python package, allowing users to `pip install sesmg`.  You could then install a command line hook to run streamlit e.g. `sesmg run`.  Using setuptools you can detect the environment, if you need to call custom setup scripts.

In other words, your current installation instructions are too opinionated. They force users into a way of working which is not very flexible.  I would recommend that you split your installation instructions into two parts 1. installation of dependencies (CBC, postgres, geos, graphviz) and provide alternative means to do this e.g. conda, brew, manual download + install and 2. pip install sesmg

**Comment by chrklemm** [comment]
Dear @willu47,
thank you for your effort and your feedback.

Within the automatic installation, which is executed with the respective "installation" file, a virtual environment is automatically set up. We decided to use an installation procedure that makes the SESMG usable without any programming knowledge. This means that no advanced IDEs are required and we do not use terminal commands. We hope that this will provide a low-threshold entry, so that users without programming and modeling knowledge can also use the SESMG.

However, we will discuss your recommendation in the dev-team and look for a solution that is convenient for all types of users.

By the way: We saw that you used Python 3.10, which support is currently not guaranteed. We recommend to use Python 3.7, 3.8 or 3.9 (see https://spreadsheet-energy-system-model-generator.readthedocs.io/en/latest/02.01.00_installation.html#macos ).

Best regards,
SESMG-Team

**Comment by GregorBecker** [comment]
Hello @willu45,

first of all we would like to thank you again for your comments regarding our installation process. We agree with you, that it was very specific and therefore did not allow the use of the SESMG in a different virtual environment than the one we specified.

After discussing and trying to implement the packaging e.g. via PyPI in the team, we figured our that this kind of deployment is not suitable for the SESMG, due to the following reasons: 

- The SESMG is not a typical python package and is therefore not imported in the classic sense, but rather represents an application, which should therefore be seen as a closed product.

- The SESMG needs the given folder structure on Github. During the optimization process the results are stored in the result folder, for example. Those paths are fixed in the code and cannot be influenced by the user.

- PyPI only provides an installation of sub-packages which are also shared via PyPI. We are using some packages which are not shared this way, which is a third barrier to use PyPI.

After we have looked at the packaging and have determined that a PyPI packaging is not purposeful, we have decided to extend the installation description by an advanced user entry, which allows the user to use other environments (e.g. Anaconda) as shown in your code section. We updated the  documentation about this additional way to install the SESMG, which you can find here: https://spreadsheet-energy-system-model-generator.readthedocs.io/en/latest/02.01.00_installation.html#advanced

We hope to aim your ideas of an alternative installation process. Otherwise, please feel free to contact us again.

Kind regards 
SESMG Dev Team

**Comment by willu47** [comment]
Hi @GregorBecker. Thanks for the responses. Given the opportunity of the review, I would like to challenge you and the team on each of these points.

> - The SESMG is not a typical Python package and is therefore not imported in the classic sense, but rather represents an application, which should therefore be seen as a closed product.

Then why not use a more complete packaging solution?  There are various solutions such as `py2exe` and `pyinstaller` that allow you to create an installable package.  You can also bundle a Flask app, freezing all dependencies and a Python version.  This would be a more complete solution to what you have now, which seems a bit in the middle.

> - The SESMG needs the given folder structure on Github. During the optimization process the results are stored in the result folder, for example. Those paths are fixed in the code and cannot be influenced by the user.

This is a design issue that could be avoided by using the Python provisions for bundling data files in applications and having access to the system and temporary folders. 

See the [setuptools](https://setuptools.pypa.io/en/latest/userguide/datafiles.html) documentation for instructions on the former.

For the latter, you could allow the user to define a results folder in the GUI, or write results to a temporary directory using one of the inbuilt Python libraries (which work across filesystems). In general, you want to avoid hardcoding file paths.

> - PyPI only provides an installation of sub-packages which are also shared via PyPI. We are using some packages which are not shared this way, which is a third barrier to use PyPI.

You can use pip to install Python packages from any location on the internet, including Git repositories, other Python packaging indexes etc.  The issue is some of the more hard-to-install packages such as GDAL which can be a pain to install on Windows if you're not using conda. But then that is an argument for using something like miniconda which will also help manage dependency compatibility.

Another solution could be to use a virtual machine to distribute the program, such as a Docker container or vagrant.
---
In summary, there's no perfect way to package and deploy your Python application. I don't think the solution you have at the moment is optimal, given your focus on ease of use.

**Comment by GregorBecker** [comment]
Hello @willu47,

thanks for the recommendations of the python packages for creating executable applications. We decided to use the following solution:

1. We have set up our application as a Streamlit GUI in a Pyside6 environment, so that the process is terminated by closing the window and the process does not continue to run in the background unintentionally.

2. We created a Github action to generate system executables running on Pyinstaller, which on the one hand makes them available after each push in the pre-release (e. g. https://github.com/SESMG/SESMG/releases/tag/SESMG_beta) and also triggers them with each semantic release, so that each version also ships with its executables.

However, we are still facing some unexpected challenges (for example, concerning our application's structure in combination with an executable PySide6 interface) that we thought we could address more quickly. Therefore we would like to apologize for the delays and will inform you as soon as we solved the few remaining issues. You can see our development progress in the issue-188 branch (https://github.com/SESMG/SESMG/tree/issue-188). We hope to be done soon.

Kind regards
SESMG Dev Team

**Comment by willu47** [comment]
This looks like a really great development and I'm looking forward to seeing the finished result!

**Comment by willu47** [comment]
Hi @chrklemm - congratulations. I have been able to test the installation for Mac OSX and can confirm that the installation process works as documented. It's a smooth experience.  It's a shame that you've got the license incompatibility problem, but there's not much you can do about that without changing your software stack.

I appreciate that this has been a longer-than-expected development process, but I think that it better meets your aims of providing a smooth user-facing experience for those who are not yet comfortable dealing with management of Python environments and batch or shell scripts.

Will

**Comment by chrklemm** [comment]
Dear @willu47,

thank you very much for your positive feedback. I have merged the changes into the master branch and I will close this issue.

Best regards,
@chrklemm
